python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
-        results = self.last_response.query_results
-        for item, dest in [(f, dest_dir.joinpath(f)) for f in (results['files'] if 'files' in results else [])]:
-            dest.parent.mkdir(exist_ok=True)
-            success = success and await self.download_from_dataset(dataset_name=dataset_name, item_name=item, dest=dest)
-        return success
-
-    async def download_from_dataset(self, dataset_name: str, item_name: str, dest: Path) -> bool:
+        req_params = {'action': action, 'dataset_name': dataset_name, 'data_location': item_name}
+
+        if self.uses_auth:
+            # This will be replaced as soon as we call apply_auth, but some string is required for __init__
+            req_params['session_secret'] = ''
+            request = MaaSDatasetManagementMessage(**req_params)
+            # TODO: (later) implement and use DmodAuthenticationFailure, though possibly down the apply_auth call stack
+            if not await self._auth_client.apply_auth(request):
+                msg = f'{self.__class__.__name__} could not apply auth to {request.__class__.__name__}'
+                raise DmodRuntimeError(msg)
+            else:
+                return request, MaaSDatasetManagementResponse
+        else:
+            return DatasetManagementMessage(**req_params), DatasetManagementResponse
+
+    async def download_dataset_item(self, dataset_name: str, item_name: str, dest: Path) -> DatasetManagementResponse:
         if dest.exists():
-            return False
+            reason = f'Destination File Exists'
+            msg = f'{self.__class__.__name__} could not download dataset item to existing path {str(dest)}'
+            return DatasetManagementResponse(success=False, reason=reason, message=msg)
+
         try:
             dest.parent.mkdir(parents=True, exist_ok=True)
         except:
-            return False
-        request = DatasetManagementMessage(action=ManagementAction.REQUEST_DATA, dataset_name=dataset_name,
-                                           data_location=item_name)
-        self.last_response: DatasetManagementResponse = await self.async_make_request(request)
+            reason = f'Unable to Create Parent Directory'
+            msg = f'{self.__class__.__name__} could not create parent directory for downloading item to {str(dest)}'
+            return DatasetManagementResponse(success=False, reason=reason, message=msg)
+
+        try:
+            request, final_response_type = self._request_prep(dataset_name=dataset_name, item_name=item_name,
+                                                              action=ManagementAction.REQUEST_DATA)
+        # TODO: (later) implement and use DmodAuthenticationFailure
+        except DmodRuntimeError as e:
+            reason = f'{self.__class__.__name__} Download Auth Failure'
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
+
         with dest.open('w') as file:
```

Comment:
I think we need to write bytes here (if we are going to handle opening the file)?

Reply:
Not sure I follow.  By the time data to write gets here, it's a `str` attribute of an inflated object.  Are you suggesting it would be a better practice to deal with files in binary mode and just cast these `str`s back to `bytes`?  Or perhaps a larger reworking that keeps the data as `bytes` somehow?

Reply:
@aaraney, pinging on this follow up question again.

Reply:
> Or perhaps a larger reworking that keeps the data as bytes somehow?

I didn't realize that `DataTransmitMessage`'s `data` field is `str`. Yeah I think we need to change the `data` field to either `Union[str, bytes]` or just `bytes`. Without accepting data as `bytes` we will get `UnicodeDecodeError` exceptions for any non-unicode data. 

For writing to the file, we could either cast everything we get to `bytes` or do a type check if it is `str` or `bytes` and open the file in the correct mode accordingly.

Reply:
As mentioned elsewhere, `data` is a `str` in order to support serialization of the messages.

Reply:
Not to open a can of worms, but as mentioned elsewhere, does this effectively limit us to uploading and downloading unicode encoded data?

Reply:
I modified #486 a bit to involve a large process related to this, though as I mentioned elsewhere, I think we kick this can down the road for now.  We can work around it if necessary, and we need to prioritize other tasks.